// restkit/auth/auth.js
"use strict";
const injector_1 = require('../injector');
const route_1 = require('../route');
const manager_1 = require('./manager');
function Auth(name) {
    return function (object, method, index) {
        let injectable = {
            index: index,
            arguments: [name]
        };
        let injectionConfig = {
            injectionResolver: new AuthenticationInjectionResolver(),
            injectable: injectable
        };
        injector_1.InjectorService.registerInjection(object, method, injectionConfig);
    };
}
exports.Auth = Auth;
class AuthenticationInjectionResolver {
    resolve(injectable, request) {
        return new Promise((resolve, reject) => {
            let name = injectable.arguments[0];
            let authResource;
            if (name) {
                authResource = manager_1.AuthManager.getHandlerByName(name);
            }
            else {
                authResource = manager_1.AuthManager.getDefault();
            }
            injector_1.InjectorService.run(authResource.object, authResource.method, request).then((response) => {
                if (response.type === 1) {
                    reject(response);
                }
                else {
                    resolve(response);
                }
            })
                .catch((response) => {
                if (!(response instanceof route_1.Response) && !(response instanceof Error)) {
                    reject(new route_1.Response(401, response));
                }
                else {
                    reject(response);
                }
            });
        });
    }
}
exports.AuthenticationInjectionResolver = AuthenticationInjectionResolver;
